Allow Zed to run under multiple user accounts simultaneously

samsonjs opened the pull request:
Closes #4607

This is an attempt to enable Zed to run under multiple user accounts on the same Mac, because it's a blocker to me really giving Zed a fair shot at being my primary editor.

According to some helpful info from @ForLoveOfCats in #4607 the main reason why this doesn't work is because Zed is using a Unix socket or maybe a TCP socket with a hard-coded path and/or port. To me it looks like it's a TCP socket so I tried changing that code in here, but I'm stuck at trying to test it out because running `target/debug/zed` or `target/release/zed` seems to behave differently than running an actual app bundle. I had no luck copying the binary over to /Applications/Zed.app/Contents/MacOS/zed because it can't find WebRTC.framework which resides at a different relative path in the app bundle.

If this seems like a desirable change to the core team then I'm looking for some guidance on how to build an app bundle or otherwise test out this change, or a nudge in the correct direction if I'm way off base with my current approach.

Release Notes:

- Added multiuser support for up to 100 users on the same machine.

samsonjs committed:
First take at running under multiple user accounts

samsonjs commented:
@cla-bot check

samsonjs commented:
Also I've never written a line of Rust before in my life so my feelings won't be hurt if you have any feedback about basic stuff in the code. Please feel free to critique anything in here to your heart's content.

osiewicz commented:
> I had no luck copying the binary over to /Applications/Zed.app/Contents/MacOS/zed because it can't find WebRTC.framework which resides at a different relative path in the app bundle.

You can use `scripts/bundle-mac -li` to "simulate" installing a dev build as a release one. You'll have a `Zed Dev.app` in your /Applications then.

osiewicz commented on crates/zed/src/main.rs:
```diff
@@ -312,7 +312,7 @@ fn main() {
     {
         if env::var("ZED_STATELESS").is_err() {
             if crate::zed::listen_for_cli_connections(open_listener.clone()).is_err() {
-                println!("zed is already running");
+                println!("zed is already running {:?}", crate::zed::listen_for_cli_connections(open_listener.clone()));
```
This is slightly wrong; you're calling this function the second time to get the error value, but it's not guaranteed that it'll fail again (even though it most likely will). Plus, `listen_for_cli_connections` has side effects.
I think it'd be better to just store the error and print it:
```
            if let Err(connection_error) = crate::zed::listen_for_cli_connections(open_listener.clone()) {
                println!("zed is already running: {connection_error:?}");
                return;
            }
```

osiewicz commented on crates/zed/src/zed/only_instance.rs:
```diff
@@ -13,14 +15,21 @@ const RECEIVE_TIMEOUT: Duration = Duration::from_millis(35);
+        ReleaseChannel::Stable => 43757,
+        ReleaseChannel::Nightly => 43767,
     };
+    let user_id = Uid::current().as_raw() as u16;
```
Are user UIDs guaranteed to be consecutive?

samsonjs replied:
Good point, they are definitely not always consecutive. There might be gaps so it's safer to have a bigger buffer in between ports for each  release channel.

mikayla-maki commented on crates/zed/src/zed/only_instance.rs:
```diff
@@ -13,14 +15,21 @@ const RECEIVE_TIMEOUT: Duration = Duration::from_millis(35);
-        ReleaseChannel::Preview => 43738,
-        ReleaseChannel::Stable => 43739,
-        ReleaseChannel::Nightly => 43740,
+        ReleaseChannel::Preview => 43747,
```
We don't want to change this.

samsonjs commented:
@osiewicz Thank you very much for the pointers! I'll see what I can come up with now

samsonjs committed:
Increase port spacing and remove debugging prints

Also swaps out the unistd call for the users crate to make this build on
Windows as well.

maxdeviant commented on crates/zed/src/zed/only_instance.rs:
```diff
@@ -13,14 +15,28 @@ const RECEIVE_TIMEOUT: Duration = Duration::from_millis(35);
-        ReleaseChannel::Preview => 43738,
-        ReleaseChannel::Stable => 43739,
-        ReleaseChannel::Nightly => 43740,
+        ReleaseChannel::Preview => 43837,
```
Maybe we could make this more obvious by performing addition instead of encoding it into a literal?

For example:

```suggestion
        ReleaseChannel::Preview => 438737 + USER_BLOCK,
```

Where `USER_BLOCK` is `100`.

samsonjs replied:
Great idea, that makes it much clearer.

samsonjs commented:
I was able to properly test this out by commenting out the check that treats all dev builds as the only instance and I think it's sound now. Rather than using unistd I swapped in the [users crate](https://crates.io/crates/users) to fix the failing Windows build.

I'm not certain about the addition overflowing behaviour in here and maybe it's worth having a check for that, any advice on that is welcome. It seems like a real corner case but I'm less certain about Windows than Unix on this one.

samsonjs committed:
PR feedback: clarify code by adding a constant

samsonjs commented:
Ok so this users crate doesn't work on Windows. I'm away right now but I have a Windows PC at home so when I'm back I can try to find some time to figure this out.

JunkuiZhang commented:
Hi, I implemented single instance on Windows using a different approach #15371 . Any advice or suggestions?

osiewicz commented:
Hey, how about using something like `sysinfo` crate?

samsonjs committed:
Remove users crate

samsonjs commented:
Thanks @osiewicz that looks like the ticket! I'm able to grab the user's `Uid` from the process however I'm stuck trying to get a `u32` value from it. This is probably basic Rust stuff so please bear with me. [The `Uid` docs say it implements `Deref`](https://docs.rs/sysinfo/0.31.3/sysinfo/struct.Uid.html#impl-Deref-for-Uid) with a `u32` target but `*uid` doesn't seem to give me a `u32` value.

![CleanShot 2024-08-23 at 12 44 14@2x](https://github.com/user-attachments/assets/0236864a-e145-4403-9e88-3105cf0af9df)

Any advice is appreciated and I'll try to find a bit of time to figure it out on my own as well.

notpeter commented:
> I'm stuck trying to get a u32 value from it

Also a rust novice, but I think `*uid.clone()` should do it.

<img width="280" alt="Screenshot 2024-08-23 at 16 40 14" src="https://github.com/user-attachments/assets/2a17ebdc-9ba8-407b-9a19-29b570820d4b">

samsonjs committed:
Try using sysinfo to get the user ID and ensure port doesn't overflow

samsonjs commented:
Yep that does the trick, thank you!

samsonjs committed:
Code style

samsonjs committed:
Extract a u32 RID from the SID on Windows

samsonjs committed:
Fix get_uid_as_u32 on Windows

samsonjs commented:
Ok I got my hands on a Windows box and I think this is finally gonna work 🤞

mikayla-maki committed:
Merge branch 'main' into 4607-multiuser

osiewicz commented:
Thanks!

samsonjs commented:
Thank you all for the assistance! I like Zed a lot so far!
